protocols/autonatv2/src/server/handler/dial_request.rs

```diff
@@ -0,0 +1,277 @@
+        Self {
+            observed_multiaddr,
+            dial_back_cmd_sender,
+            dial_back_cmd_receiver,
+            inbound: FuturesSet::new(Duration::from_secs(10), 10),
+            rng,
+        }
+    }
+}
+
+impl<R> ConnectionHandler for Handler<R>
+where
+    R: RngCore + Send + Clone + 'static,
+{
+    type FromBehaviour = ();
+
+    type ToBehaviour = io::Result<DialBackCommand>;
+
+    type InboundProtocol = ReadyUpgrade<StreamProtocol>;
+
+    type OutboundProtocol = DeniedUpgrade;
+
+    type InboundOpenInfo = ();
+
+    type OutboundOpenInfo = ();
+
+    fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
+        SubstreamProtocol::new(REQUEST_UPGRADE, ())
+    }
+
+    fn poll(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<
+        ConnectionHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::ToBehaviour>,
+    > {
+        match self.inbound.poll_unpin(cx) {
+            Poll::Ready(Ok(Err(e))) => {
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Err(e)));
+            }
+            Poll::Ready(Err(e)) => {
+                return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Err(
+                    io::Error::new(io::ErrorKind::TimedOut, e),
+                )));
+            }
+            Poll::Ready(Ok(Ok(_))) => {}
+            Poll::Pending => {}
+        }
+        if let Poll::Ready(Some(cmd)) = self.dial_back_cmd_receiver.poll_next_unpin(cx) {
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Ok(cmd)));
+        }
+        Poll::Pending
+    }
+
+    fn on_behaviour_event(&mut self, _event: Self::FromBehaviour) {}
+
+    fn on_connection_event(
+        &mut self,
+        event: ConnectionEvent<
+            Self::InboundProtocol,
+            Self::OutboundProtocol,
+            Self::InboundOpenInfo,
+            Self::OutboundOpenInfo,
+        >,
+    ) {
+        match event {
+            ConnectionEvent::FullyNegotiatedInbound(FullyNegotiatedInbound {
+                protocol, ..
+            }) => {
+                if self
+                    .inbound
+                    .try_push(handle_request(
+                        protocol,
+                        self.observed_multiaddr.clone(),
+                        self.dial_back_cmd_sender.clone(),
+                        self.rng.clone(),
+                    ))
+                    .is_err()
+                {
+                    tracing::warn!(
+                        "failed to push inbound request handler, too many requests in flight"
+                    );
+                }
+            }
+            ConnectionEvent::ListenUpgradeError(ListenUpgradeError { error, .. }) => {
+                tracing::debug!("inbound request failed: {:?}", error);
+            }
+            _ => {}
+        }
+    }
+
+    fn connection_keep_alive(&self) -> bool {
+        true
+    }
```

Comment:
This will keep the connection alive _beyond_ any streams that are active. I don't think that should be the case.

Reply:
Yes, you are right.